Fix minimizing to tray when requested by window manager

kempniu opened the pull request:
Hi Todd,

First of all, thanks for your work on this application!

I'm skimming through the commit history and I keep wondering why have you abandoned support for minimizing the window to tray when it is requested by the window manager (first implemented in e132f7c). Introducing the change from this pull request made it work for me (in XFCE 4.12) and didn't break "normal" minimization when `--hide-via-tray` is not supplied. Clicking on the tray icon also results in the desired behavior in both minimization modes.

What do you think? Am I missing something here?

kempniu committed:
Fix minimizing to tray when requested by window manager

twolfson commented:
I believe I played around with that in nw.js for `plaidchat` (i.e. triggering a hide instead of a minimize). Instead of stopping the normal action, it wound up running both which was useless =/

I will have to play around with what you have to verify that it does/doesn't stop the default minimize mechanism. Also, if we wind up supporting this PR, then this option should be `minimize-to-tray`. I chose the `hide-via-tray` naming because it implies that only the tray controls hiding/showing.

kempniu commented:
As you probably suspect, I wasn't aware of that context and thus I suggested the simplest working solution that came to my mind for this specific application. I'll update the pull request branch after the weekend so that a separate CLI switch is used for enabling minimizing to tray.

twolfson commented:
I played around with the PR and it looks like the animation on Linux is a minimize but it also does the hide as well. It looks like in all cases there is no window positioning loss (this was a problem in nw.js) so we should be good with this PR once the flags get sorted out.

twolfson commented:
To make sure we are on the same page, can we build `--minimize-to-tray` such that it builds on top of `--hide-via-tray` (e.g. we use `program.hideViaTray || program.minimizeToTray` in option checks)?

kempniu commented:
Erm, not sure what you mean by "building on top of" `--hide-via-tray`. After reading your [first comment](#issuecomment-120457561), I understood you want the following:

| CLI switch(es) | Action: tray icon | Action: minimize button |
| --- | --- | --- |
| `--hide-via-tray` | hide or show | minimize |
| `--minimize-to-tray` | minimize or show | hide |
| `--hide-via-tray --minimize-to-tray` | hide or show | hide |

But then you mentioned a logical alternative in "option checks" (do you mean the conditional statement in b3a6a65?) and now I'm confused :)

twolfson commented:
Sorry, let me rephrase. I think that as a user I would expect `--minimize-to-tray` to  automatically pick up the behaviors that `--hide-via-tray` has (i.e. when I click on the tray icon, it should trigger a window hide). As a result, all conditional statements that were initially for `program.hideViaTray` should be updated to `program.hideViaTray || program.minimizeToTray`.

kempniu commented:
Ah, OK, it's clear now and I agree. Will update the PR tomorrow.

kempniu committed:
Add --minimize-to-tray option

New command line option, --minimize-to-tray, has been added. This allows
the user to choose whether the window should be hidden to tray only when
clicking the tray icon (--hide-to-tray) or also when minimizing the
window using the window manager (--minimize-to-tray).

kempniu commented:
I added the new option and also decided to change the help description for `--hide-to-tray` as IMHO things would get confusing if I didn't. On my box, both options work as expected. Nothing bad also happens when you use them both at the same time. Make sure to let me know if you have any comments.

twolfson commented on lib/google-music-electron.js:
```diff
@@ -15,7 +15,8 @@ var pkg = require('../package.json');
   .option('-S, --skip-taskbar', 'Skip showing the application in the taskbar')
-  .option('--hide-via-tray', 'Replace default minimize action with hiding the window')
+  .option('--minimize-to-tray', 'Hide window to tray instead of minimizing')
+  .option('--hide-via-tray', 'Hide window to tray instead of minimizing (only for tray icon)')
```
:+1: Clarification looks good

twolfson commented:
This PR looks good. I will land it by the end of the weekend.

twolfson commented:
This has been merged/released in `1.13.0`. Thanks for the quick implementation and help on this =)

Would you be interested in becoming a collaborator on this project? It would effectively mean that I start running PRs by you before landing them.

kempniu commented:
I appreciate the offer, I really do, but I don't want to make promises I might not be able to keep due to my current responsibilities. I'll be using this project daily, though, so you can count on me suggesting fixes and improvements if I come up with any :)

twolfson commented:
No worries. If you ever change your mind, let me know

kempniu commented:
Will do, thanks.
